File: challenges/222/python/solution.py
#!/usr/bin/env python


class Solution:
    def countNodes(self, root, l=1, r=1):

        if not root:
            return 0

        left = right = root
        while left := left.left:
            l += 1
        while right := right.right:
            r += 1

        if l == r:
            return 2**l - 1  # if it's a full tree, its size is known

        return 1 + self.countNodes(root.left) + self.countNodes(root.right)


if __name__ == "__main__":
    pass

File: challenges/222/js/solution.ts
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

export const countNodes = (root, l = 1, r = 1) => {
  if (!root) {
    return 0;
  }

  let left = root,
    right = root;

  while ((left = left.left)) {
    l += 1;
  }
  while ((right = right.right)) {
    r += 1;
  }

  if (l == r) {
    return Math.pow(2, l) - 1;
  }

  return 1 + countNodes(root.left) + countNodes(root.right);
};
